capstone/src/main/java/com/google/sps/servlets/UserGroupsServlet.java

```diff
@@ -0,0 +1,93 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.sps.Objects.User;
+import com.google.sps.Objects.Group;
+import com.google.sps.Objects.Badge;
+import com.google.sps.Objects.response.UserGroupResponse;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import java.io.IOException;
+import java.io.PrintWriter;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.LinkedHashSet;
+import com.google.gson.Gson;
+import error.ErrorHandler;
+
+/**
+ * Servlet that returns a user's groups data from the Datastore.
+ */
+@WebServlet("/user-groups")
+public class UserGroupsServlet extends HttpServlet {
+
+  /**
+   * Gets User's Group data from the Datastore and returns it.
+   */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    Entity currentUser = null;
+    ArrayList<UserGroupResponse> groups = new ArrayList<>();
+
+    if (userService.isUserLoggedIn()) {
+      String userId = userService.getCurrentUser().getUserId();
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      Key entityKey = KeyFactory.createKey("User", userId);
+      try {
+        currentUser = datastore.get(entityKey);
+      } catch (EntityNotFoundException e) {
+        ErrorHandler.sendError(response, "User not found.");
+        return;
```

Comment:
UserDataServlet still returns null on doGet, whereas here you `return;` early. I don't know which is correct, but I would find a consistent solution for error handling and what gets returned to the server. Have you been able to test these methods with some fake values to see what happens in these error cases?

Reply:
Hmm this led me to do some testing in the error cases to see exactly what happens. The error is correctly thrown and the error message html is displayed at the server url. But our ErrorHandler returns an html response, while the javascript parses the response as json, so in addition to the expected error message sent by the server, there's also "Unexpected token < in JSON". So I think we could modify ErrorHandler to send a json response instead, and generate the error message displayed to the user in the javascript. Or we could have the response.sendRedirect() to the servlet page. Not sure which is the better method.

Reply:
Okay, we decided that for the users it would be nice to display an alert message with the error message. This would probably be done on the client side, which raises the question of whether we really need ErrorHandler at all. I think if we return null in these cases then it would be easier to check for the error condition on the client side, but ErrorHandler itself wouldn't serve much purpose.

Reply:
Well you'd likely still want SOMETHING server side to check for an error and classify the error for the user right? If a user tries to delete a post and the error is that the post does not exist, that's going to require something on the server to check for the existence of the post and then report that error. I would suggest modifying the existing ErrorHandler sendError to to just send a json which has any error you'd want to display to the user.

Reply:
Just to clarify, is the "Unexpected token < in JSON" a JS error or something that is displayed to the user?

If you return null in each of these cases, how would you know what is the proper alert message to display to the user on the clientside? 
I had sent Lucy these links to help her set up the ErrorHandler: 
https://www.programcreek.com/java-api-examples/?class=javax.servlet.http.HttpServletResponse&method=sendError
https://stackoverflow.com/questions/3885725/setting-both-error-message-and-body-from-servlet
The way the ErrorHandler has been implemented, you don't need to do anything on the clientside because it'll just display the error message that was sent. 

While I agree that a proper clientside error handler would be useful, you'd need it to work in conjunction with the server side, something along the lines of: server side sends an error code/message and then the clientside decides what to do with the error it receives. Since we are displaying the error to the user ATM, I don't think we should focus on changing it right now but rather as a stretch feature should we have time later.

Reply:
The unexpected token in JSON was a JS error. Because the ErrorHandler was sending a response with html in it, but the JS was trying to parse it as JSON.

You're right, the issue with just sending nulls would be that you can't distinguish what the error was caused by, so I guess we would need to keep ErrorHandler. I'm pretty sure the ErrorHandler we're using is implemented based off of the links you sent. However, nothing appears on the clientside. So I think modifying ErrorHandler to send a JSON might be the best solution. I can focus on that later though if you think it's not a priority for right now.